rfcs/20200920-keras-nlp-bert.md

```diff
@@ -0,0 +1,369 @@
+```python
+# Considering a imbd review dataset
+import tensorflow as tf
+import tensorflow_datasets as tfds
+import keras_nlp
+import tensorflow_text as tftext
+
+imdb_reviews = tfds.load('imdb_reviews')
+train_ds = imdb_reviews['train'].batch(32)
+test_ds = imdb_reviews['test'].batch(32)
+
+# Tokenization with BertTokenizer
+vocab_path = "gs://<bucket_name>/<file_path>/vocab.txt"
+tokenizer = tftext.BertTokenizer(vocab_path, token_out_type=tf.int64, lower_case=False)
+MAX_SEQ_LENGTH = 128
+def preprocess(input_text):
+  token_ids = tokenizer.tokenize(input_text)
+  cls = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([101], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  sep = tf.expand_dims(tf.expand_dims(tf.cast(tf.tile([102], [token_ids.nrows()]), tf.int64), axis=1), axis=1)
+  token_ids = tf.concat([cls, token_ids, sep], axis=1)
+  segment_ids = tf.concat([tf.zeros_like(cls), tf.ones_like(token_ids), tf.ones_like(sep)], axis=1)
+  output_shape = [None, SEQUENCE_LENGTH]
+  token_ids = token_ids.merge_dims(-2, -1)
+  segment_ids = segment_ids.merge_dims(-2, -1).to_tensor(shape=output_shape)
+  input_mask = tf.ones_like(token_ids).to_tensor(shape=output_shape)
+  token_ids = token_ids.to_tensor(shape=output_shape)
+  return {
+      'input_ids': token_ids,
+      'input_mask': input_mask,
+      'input_type_ids': segment_ids
+  }
+
+strategy = tf.distribute.TPUStrategy(...)
+with strategy.scope():
+  encoder = keras_nlp.encoders.BertEncoder(vocab_size=30522, max_sequence_length=512, type_vocab_size=2)
+  checkpoint_path = "gs://<bucket_name>/<file_path>"
+  checkpoint = tf.train.Checkpoint(model=encoder)
+  checkpoint.restore(checkpoint_path).assert_consumed().run_restore_ops()
+  input_ids = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_ids', dtype=tf.int32)
+  input_mask = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_mask', dtype=tf.int32)
+  label_ids = tf.keras.layers.Input(shape=(128,), name='input_type_ids', dtype=tf.int32)
+  x = encoder([input_ids, input_mask, label_ids])['pooled_output']
+  x = tf.keras.layers.Dropout(rate=0.1)(x)
+  output = tf.keras.layers.Dense(1, activation='sigmoid')(x)
+  model = tf.keras.Model(inputs=[input_ids, input_mask, label_ids], outputs=output)
+
+model.compile('adam', 'binary_crossentropy', ['accuracy'])
+model.fit(train_ds, epochs=5, validation_data=test_ds)
+```
+
+### Pretraining task
+
+We aim to provide SavedModel for BertEncoder, however user can choose to pretrain a new BertEncoder based on their
+own dataset.
+
+```python
+with strategy.scope():
+  encoder = keras_nlp.encoders.BertEncoder(vocab_size, max_sequence_length, type_vocab_size)
+  input_ids = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_ids', dtype=tf.int32)
+  input_mask = tf.keras.layers.Input(shape=(SEQUENCE_LENGTH,), name='input_mask', dtype=tf.int32)
+  label_ids = tf.keras.layers.Input(shape=(128,), name='input_type_ids', dtype=tf.int32)
```

Comment:
If this is the mask token ids, it is not used inside this keras model?
How to use inside the loss?

Reply:
Yeah renamed this to type_id instead.